utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/encode/AvroSchemaOutputValidator.scala

```diff
@@ -0,0 +1,236 @@
+    val schemaFieldsValidation = map.fields.map{ case (key, value) =>
+      val fieldPath = buildPath(key, path)
+      validateTypingResult(value, schema.getValueType, fieldPath)(isNullable = false)
+    }.foldLeft[ValidatedNel[OutputValidatorError, Unit]](().validNel)((a, b) => a combine b)
+
+    schemaFieldsValidation
+  }
+
+  private def validateArraySchema(array: TypedClass, schema: Schema, path: Option[String])(implicit isNullable: Boolean) = {
+    def validateListElements(params: List[TypingResult]) = {
+      params
+        .distinct
+        .zipWithIndex
+        .map { case (el, index) =>
+          val pathKey = params match {
+            case _ :: Nil => "" //One element list is 'general' object, and we preset it as field[]
+            case _ => index.toString
+          }
+
+          val elementPath = buildPath(pathKey, path, isArray = true)
+          validateTypingResult(el, schema.getElementType, elementPath)(isNullable = false)
+        }
+        .foldLeft[ValidatedNel[OutputValidatorError, Unit]](().validNel)((a, b) => a combine b)
+    }
+
+    //Empty array is presented as List(Unknown) or List()
+    val isEmptyArray = array.params match {
+      case head :: _ => head match {
+        case Unknown => true
+        case _ => false
+      }
+      case Nil => true
+      case _ => false
+    }
+
+    if (isEmptyArray) {
+      canBeSubclassOf(array, schema, path)
+    } else {
+      validateListElements(array.params)
+    }
+
+  }
+
+  private def validateUnionSchema(union: TypingResult, schema: Schema, path: Option[String]) = {
+    val nestedObjectsInUnion = schema.getTypes.asScala.exists(sch => nestedObjects.contains(sch.getType))
+
+    val results = if (nestedObjectsInUnion) { // we want to go inside object and verify each field.. (we don't verify potential root nullability)
+      implicit val isNullable: Boolean = schema.isNullable
+      schema.getTypes.asScala.filterNot(_.isNullable).map(validateTypingResult(union, _, path)).toList
+    } else {
+      implicit val isNullable: Boolean = false
+      schema.getTypes.asScala.map(validateTypingResult(union, _, path)).toList
+    }
+
+    if (results.exists(_.isValid)) {
+      valid
+    } else {
+      results.foldLeft[ValidatedNel[OutputValidatorError, Unit]](().validNel)((a, b) => a combine b)
+    }
+  }
+
+  private def validateClass[T:ClassTag](typingResult: TypingResult, schema: Schema, path: Option[String]): Validated[NonEmptyList[OutputValidatorTypeError], Unit] = {
+    val clazz: Class[_] = implicitly[ClassTag[T]].runtimeClass
+    validateClass(typingResult, schema, List(clazz), path)
+  }
+
+  private def validateClass(typingResult: TypingResult, schema: Schema, allowedClasses: List[Class[_]], path: Option[String]): Validated[NonEmptyList[OutputValidatorTypeError], Unit] = {
+    val schemaAsTypedResult = AvroSchemaTypeDefinitionExtractor.typeDefinition(schema)
+
+    (schemaAsTypedResult, typingResult) match {
```

Comment:
it would be easier for me to understand if it was just
(schemaAsTypedResult :: additionalAllowedClasses).contains(typedClass)

Reply:
I thought more about additionalAllowedClasses: TypingResult etc,